internal/function/commit_has_tree.go

```diff
@@ -0,0 +1,120 @@
+package function
+
+import (
+	"io"
+
+	"github.com/src-d/gitquery"
+	git "gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/plumbing"
+	"gopkg.in/src-d/go-mysql-server.v0/sql"
+	"gopkg.in/src-d/go-mysql-server.v0/sql/expression"
+)
+
+// CommitHasTree is a function that checks whether a tree is part of a commit
+// or not.
+type CommitHasTree struct {
+	expression.BinaryExpression
+}
+
+// NewCommitHasTree creates a new CommitHasTree function.
+func NewCommitHasTree(commit, tree sql.Expression) sql.Expression {
+	return &CommitHasTree{expression.BinaryExpression{
+		Left:  commit,
+		Right: tree,
+	}}
+}
+
+// Name implements the Expression interface.
+func (CommitHasTree) Name() string { return "commit_has_tree" }
+
+// Type implements the Expression interface.
+func (CommitHasTree) Type() sql.Type { return sql.Boolean }
+
+// Eval implements the Expression interface.
+func (f *CommitHasTree) Eval(s sql.Session, row sql.Row) (interface{}, error) {
+	session, ok := s.(*gitquery.Session)
+	if !ok {
+		return nil, gitquery.ErrInvalidGitQuerySession.New(s)
+	}
+
+	left, err := f.Left.Eval(s, row)
+	if err != nil {
+		return nil, err
+	}
+
+	if left == nil {
+		return nil, nil
+	}
+
+	left, err = sql.Text.Convert(left)
+	if err != nil {
+		return nil, err
+	}
+
+	right, err := f.Right.Eval(s, row)
+	if err != nil {
+		return nil, err
+	}
+
+	if right == nil {
+		return nil, nil
+	}
+
+	right, err = sql.Text.Convert(right)
+	if err != nil {
+		return nil, err
+	}
+
+	commitHash := plumbing.NewHash(left.(string))
+	treeHash := plumbing.NewHash(right.(string))
+
+	iter, err := session.Pool.RepoIter()
+	if err != nil {
+		return nil, err
+	}
+
+	for {
+		repo, err := iter.Next()
+		if err == io.EOF {
+			return false, nil
+		}
+
+		if err != nil {
+			return nil, err
+		}
+
+		ok, err := f.commitHasTree(repo.Repo, commitHash, treeHash)
+		if err == plumbing.ErrObjectNotFound {
+			continue
+		}
+
+		return ok, nil
+	}
+}
+
+func (f *CommitHasTree) commitHasTree(
+	repo *git.Repository,
+	commitHash, treeHash plumbing.Hash,
+) (bool, error) {
+	commit, err := repo.CommitObject(commitHash)
+	if err != nil {
+		return false, err
+	}
+
+	return commit.TreeHash == treeHash, nil
```

Comment:
We should check all the tree objects, not only the root one.

Reply:
Only other thing related to trees `*object.Commit` has is `Tree`, which returns the tree pointed to by `TreeHash`, and then the only thing this has is the `TreeEntries`. Do you mean that? Wouldn't that be a tree entry and not a tree?